core/DataPrefetcher.cpp

```diff
@@ -0,0 +1,121 @@
+        setEngine(std::move(prefetch_engine));
+
+        prefetcher_queue_credits_in_.registerConsumerHandler(CREATE_SPARTA_HANDLER_WITH_DATA(
+            DataPrefetcher, receivePrefetchQueueCredits_, uint32_t));
+
+        sparta::StartupEvent(node, CREATE_SPARTA_HANDLER(DataPrefetcher, sendInitialCredits_));
+    }
+
+    // Send fetch the initial credit count
+    void DataPrefetcher::sendInitialCredits_()
+    {
+        req_queue_credits_out_.send(req_queue_.capacity());
+    }
+
+    //////////////////////////////////////////////////////////////////////
+    // callbacks
+
+    //! Process the incoming memory access and send to the engine
+    void
+    DataPrefetcher::processIncomingReq(const olympia::MemoryAccessInfoPtr & mem_access_info_ptr)
+    {
+        // queue in incoming buffer
+        req_queue_.push(mem_access_info_ptr);
+        ev_handle_incoming_req_.schedule(sparta::Clock::Cycle(0));
+    }
+
+    //! handler for incoming requests
+    void DataPrefetcher::handleIncomingReq_()
+    {
+        auto access = req_queue_.read(0);
+        req_queue_.pop();
+        // send a credit back
+        req_queue_credits_out_.send(1);
+
+        handleMemoryAccess(access);
+
+        if (!req_queue_.empty())
+        {
+            ev_handle_incoming_req_.schedule(sparta::Clock::Cycle(1));
+        }
+        if (prefetcher_credits_ > 0)
+        {
+            ev_gen_prefetch_.schedule(sparta::Clock::Cycle(1));
+        }
+        return;
+    }
+
+    //! \brief Recieve prefetcher queue Credits
+    void DataPrefetcher::receivePrefetchQueueCredits_(const uint32_t & credits)
+    {
+        prefetcher_credits_ += credits;
+        if (getPrefetchEngine()->isPrefetchReady())
+        {
+            // cancel any future possible event
+            ev_gen_prefetch_.cancel();
+            // generate the prefetch in current cycle
+            ev_gen_prefetch_.schedule(sparta::Clock::Cycle(0));
+        }
+    }
+
+    //! \brief Flush handler
+    void DataPrefetcher::handleFlush(const olympia::FlushManager::FlushingCriteria & criteria)
+    {
+        req_queue_credits_out_.send(req_queue_.size());
+        req_queue_.clear();
+
+        ev_gen_prefetch_.cancel();
+        ev_handle_incoming_req_.cancel();
+    }
+
+    //! handle to generate a prefetch
+    void DataPrefetcher::generatePrefetch_()
+    {
+        if (getPrefetchEngine()->isPrefetchReady())
+        {
+            // get next prefetch from the engine
+            auto access = getPrefetchEngine()->getPrefetchMemoryAccess();
+            getPrefetchEngine()->popPrefetchMemoryAccess();
+
+            // send access to output port and consume credit
+            sendPrefetch(access);
+            prefetcher_credits_--;
+
+            // schedule next prefetch generation (if any)
+            if (getPrefetchEngine()->isPrefetchReady() && (prefetcher_credits_ > 0))
+            {
+                ev_gen_prefetch_.schedule(sparta::Clock::Cycle(1));
+            }
+        }
```

Comment:
Actually, I just realize that I already checked for isPrefetchReady(), will remove this part of conditional. 
Here I am scheduling the next prefetch in queue (if I have credits)

Reply:
Scratch my earlier comment. This check is required as we have already handled one prefetch this cycle. However, there may or may not be pending prefetches already in queue which should be taken care of in subsequent cycles. 

There is no else, because this event will be scheduled in both cases of new prefetch requests being ready and prefetch credits coming back from the consuming unit